File: mobile/app/src/main/java/edu/rellon/AudioRent/common/api/ApiService.kt
package edu.rellon.audiorent.common.api

import retrofit2.Call
import retrofit2.http.*

interface ApiService {
    @POST("auth/login")
    fun login(@Body request: LoginRequest): Call<LoginResponse>

    @POST("auth/google")
    fun googleAuth(@Body request: Map<String, String>): Call<LoginResponse>

    @POST("auth/register")
    fun register(@Body request: RegisterRequest): Call<UserResponse>

    @GET("packages")
    fun getPackages(): Call<List<PackageResponse>>

    @GET("packages/{id}")
    fun getPackageById(@Path("id") id: String): Call<PackageResponse>

    @POST("rentals")
    fun createRental(@Header("Authorization") token: String, @Body request: RentalRequest): Call<RentalResponse>

    @GET("rentals")
    fun getMyRentals(@Header("Authorization") token: String): Call<List<RentalResponse>>

    @GET("admin/users")
    fun getAllUsers(@Header("Authorization") token: String): Call<List<UserResponse>>

    @PUT("admin/users/{userId}/deactivate")
    fun deactivateUser(@Header("Authorization") token: String, @Path("userId") userId: String): Call<Void>

    @GET("admin/packages")
    fun getAllAdminPackages(@Header("Authorization") token: String): Call<List<PackageResponse>>

    @GET("packages/my")
    fun getMyProviderPackages(@Header("Authorization") token: String): Call<List<PackageResponse>>

    @Multipart
    @POST("packages")
    fun createProviderPackage(
        @Header("Authorization") token: String,
        @Part("name") name: okhttp3.RequestBody,
        @Part("description") description: okhttp3.RequestBody,
        @Part("price") price: okhttp3.RequestBody,
        @Part("quantity") quantity: okhttp3.RequestBody,
        @Part("category") category: okhttp3.RequestBody,
        @Part images: List<okhttp3.MultipartBody.Part>?
    ): Call<PackageResponse>

    @GET("rentals/provider")
    fun getProviderRentals(@Header("Authorization") token: String): Call<List<RentalResponse>>

    @DELETE("admin/packages/{packageId}")
    fun deletePackage(@Header("Authorization") token: String, @Path("packageId") packageId: String): Call<Void>
}

File: web/src/common/ApiService.ts
import axios, { AxiosInstance } from 'axios';

class ApiService {
    private static instance: ApiService;
    private axiosInstance: AxiosInstance;

    private constructor() {
        // Singleton pattern: private constructor prevents instantiation
        this.axiosInstance = axios.create({
            baseURL: process.env.REACT_APP_API_BASE_URL || "http://localhost:8080",
        });

        this.axiosInstance.interceptors.request.use((config) => {
            const token = localStorage.getItem("token");
            if (token) {
                config.headers.Authorization = `Bearer ${token}`;
            }
            return config;
        });
    }

    public static getInstance(): ApiService {
        if (!ApiService.instance) {
            ApiService.instance = new ApiService();
        }
        return ApiService.instance;
    }

    public getAxiosInstance(): AxiosInstance {
        return this.axiosInstance;
    }

    // --- Admin User Management ---
    public async getAllUsers() {
        return this.axiosInstance.get("/admin/users");
    }

    public async deactivateUser(userId: string) {
        return this.axiosInstance.put(`/admin/users/${userId}/deactivate`);
    }

    // --- Admin Package Management ---
    public async getAllAdminPackages() {
        return this.axiosInstance.get("/admin/packages");
    }

    public async deletePackage(packageId: string) {
        return this.axiosInstance.delete(`/admin/packages/${packageId}`);
    }

    // --- Provider Management ---
    public async getProviderRentals() {
        return this.axiosInstance.get("/rentals/provider");
    }

    public async updateRentalStatus(rentalId: string, status: string) {
        return this.axiosInstance.put(`/rentals/${rentalId}/status?status=${status}`);
    }

    // --- Payment Management ---
    public async processPayment(rentalId: string) {
        return this.axiosInstance.post("/api/v1/payments", { rentalId });
    }
}

export default ApiService;
